protocol/test-infra.md

```diff
@@ -0,0 +1,358 @@
+  after spinning up the network, we still need to run tests against it.
+
+This might get more approachable with more separation between node setup and testing,
+as proposed in the feature-growth section.
+Ideally we then run common tests against different combinations of each of the client implementations.
+
+### Growth of activity
+
+With more vertical scaling comes more concern around performance and stability.
+We need more benchmarks and uptime checks to understand where we can improve.
+
+Today:
+- [op-ufm (user facing monitoring)](https://github.com/ethereum-optimism/infra/tree/main/op-ufm)
+- [`replayor`](https://github.com/danyalprout/replayor)
+- Internal infra dashboards / alerts
+- One-off chain activity analysis
+
+Challenges
+- `op-ufm` should run against more chains, and be more visible to engineers.
+  E.g. if transaction inclusion speed is not good, we need to do something about it.
+- `replayor` needs automation, ideally on top of a shadow-fork of a real network,
+  such that we can analyze performance without spending real ETH or harming a production network in any way.
+- We need to monitor node performance better.
+  `pprof` / Go-resource dashboards should improve and be more center to our work.
+
+Automating flight-recording in `op-service` would be great.
+See https://go.dev/blog/execution-traces-2024 for information about flight-recording.
+E.g. on RPC call or on particular pre-programmed conditions,
+a buffer of performance data from the last N seconds can be dumped and uploaded to some service.
+When slow blocks occur, or engineers are interested in a performance snapshot of a real network, this could be great.
+Also see [op-geth flight-recording design-doc (internal)](https://github.com/ethereum-optimism/design-docs-private/blob/main/op-geth-flight-recording.md).
+
+There also exists https://pyroscope.io/ for nice continuous monitoring,
+although it is built on top of older Golang profiling hooks, and seems to provide less fine-grained information.
+
+### Growth as platform
+
+The more the stack becomes a "platform" that others fork and build new things on top on,
+the more thought we should give to the public interface and extensibility of our testing.
+
+Today: no testing platform.
+
+Challenges:
+External OP Stacks forks end up having to fork the testing infra as well, and wire in their customizations.
+A lot of tests may work by default, but some features may not.
+E.g. 4844 blob tests may need to be disabled for an alt-DA network test suite.
+
+There may be ways we can better categorize tests, make test setup more composable,
+and improve parametrization of contract versioning and chain configurations.
+
+### Interoperability
+
+Interoperability is a net-new domain, and requires some deeper work than adjustments to handle growth.
+Specifically, this requires cross-chain testing.
+This testing adds multi-L2 deployments, multi-L2 test setup, and multi-L2 tests to the testing scope.
+
+Today:
```

Comment:
I'd love to find a way to reduce the amount of duplication here so that we can use a shared platform.